File: apps/api/src/modules/idp/signature.ts
import { createHmac, timingSafeEqual } from "node:crypto";

// lazyfga-21: 설정형 webhook 서명 검증 엔진. provider별 코드 대신 WebhookSignatureSpec 하나로
// ZITADEL / Stripe-style / Standard Webhooks / GitHub 등을 모두 검증한다(순수, node:crypto).

export interface WebhookSignatureSpec {
  /** 서명 헤더 이름. 예: "ZITADEL-Signature", "webhook-signature". */
  header: string;
  /** 헤더 파싱 형식. */
  headerFormat: "kv_t_v" | "scheme_hex" | "standard_webhooks";
  /** 타임스탬프 출처. */
  timestampSource: "signature_header" | "separate_header" | "none";
  /** timestampSource="separate_header"일 때 타임스탬프 헤더 이름(예: "webhook-timestamp"). */
  timestampHeader?: string;
  /** 타임스탬프 단위. "millis"는 미래 대비(현재 preset 미사용). */
  timestampUnit: "seconds" | "millis" | "none";
  /** {id}를 쓰는 템플릿용 id 출처(예: Standard Webhooks "webhook-id"). */
  idSource?: { header: string };
  /** 서명 대상 페이로드 템플릿. placeholders: {body} {timestamp} {id}. */
  payloadTemplate: string;
  algorithm: "sha256";
  /** 헤더에 실린 서명 값의 인코딩. */
  encoding: "hex" | "base64";
  /** 저장된 secret을 HMAC 키 바이트로 디코딩하는 방식. */
  secretEncoding: "raw" | "base64";
  /** 디코딩 전 제거할 접두사(예: Standard Webhooks "whsec_"). */
  secretPrefix?: string;
  /** replay 허용 윈도우(초). too-old + (의도적 하드닝) far-future 모두 거부. */
  toleranceSec: number;
  /** 다중 서명 허용(키 회전). */
  allowMultipleSignatures: boolean;
}

/** 현재 시각(초). 테스트가 주입할 수 있게 분리(결정성). */
function nowSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

/** 헤더에서 (timestamp 초, 서명값 목록)을 추출. 형식 위반 시 null. */
// 타임스탬프는 **raw 문자열**로 보존한다(서명 페이로드는 서명자가 쓴 그대로의 문자열을 써야 한다 —
// Number 라운드트립으로 "1782751299.0"·"...e0"·leading-zero를 정규화하면 검증이 어긋난다).
const DECIMAL_RE = /^[0-9]+$/;

function parseHeader(
  spec: WebhookSignatureSpec,
  headers: Headers,
): { timestampRaw?: string; signatures: string[] } | null {
  const raw = headers.get(spec.header);
  if (!raw) return null;

  let signatures: string[] = [];
  let inHeaderTsRaw: string | undefined;

  if (spec.headerFormat === "kv_t_v") {
    // "t=<ts>,v1=<sig>[,v1=<sig>]"
    const parts: Record<string, string[]> = {};
    for (const seg of raw.split(",")) {
      const i = seg.indexOf("=");
      if (i < 0) continue;
      const k = seg.slice(0, i).trim();
      const v = seg.slice(i + 1).trim();
      (parts[k] ??= []).push(v);
    }
    const tRaw = parts.t?.[0];
    if (tRaw !== undefined) {
      // 엄격한 10진 정수만(".0"/"e0"/leading-zero 거부 — ZITADEL strconv.ParseInt과 동일하게).
      if (!DECIMAL_RE.test(tRaw)) return null;
      inHeaderTsRaw = tRaw;
    }
    signatures = (parts.v1 ?? []).filter((s) => s !== "");
  } else if (spec.headerFormat === "scheme_hex") {
    // "sha256=<hex>" — scheme label은 알고리즘과 일치해야 한다(예: md5=...를 sha256으로 비교 방지).
    const eq = raw.indexOf("=");
    if (eq < 0) return null;
    if (raw.slice(0, eq).trim().toLowerCase() !== spec.algorithm) return null;
    const sig = raw.slice(eq + 1).trim();
    if (sig === "") return null;
    signatures = [sig];
  } else {
    // standard_webhooks: "v1,<base64>[ v1,<base64>]"
    for (const tok of raw.split(/\s+/)) {
      const comma = tok.indexOf(",");
      if (comma < 0) continue;
      const ver = tok.slice(0, comma);
      const sig = tok.slice(comma + 1);
      if (ver === "v1" && sig !== "") signatures.push(sig);
    }
  }

  if (signatures.length === 0) return null;
  if (!spec.allowMultipleSignatures) signatures = signatures.slice(0, 1);

  // 타임스탬프 출처 해석(raw 문자열로 보존).
  let timestampRaw: string | undefined;
  if (spec.timestampSource === "signature_header") {
    timestampRaw = inHeaderTsRaw;
    if (timestampRaw === undefined) return null;
  } else if (spec.timestampSource === "separate_header") {
    const tv = spec.timestampHeader ? headers.get(spec.timestampHeader) : null;
    if (tv === null || !DECIMAL_RE.test(tv)) return null;
    timestampRaw = tv;
  }
  return { timestampRaw, signatures };
}

/** spec.secretEncoding/secretPrefix에 따라 HMAC 키 바이트를 만든다. */
function hmacKey(spec: WebhookSignatureSpec, secret: string): Buffer {
  let s = secret;
  if (spec.secretPrefix && s.startsWith(spec.secretPrefix)) s = s.slice(spec.secretPrefix.length);
  return spec.secretEncoding === "base64" ? Buffer.from(s, "base64") : Buffer.from(s, "utf8");
}

/** 타임스탬프(초)를 toleranceSec 양방향 윈도우로 검사. */
function withinTolerance(spec: WebhookSignatureSpec, tsRaw: number): boolean {
  const ts = spec.timestampUnit === "millis" ? Math.floor(tsRaw / 1000) : tsRaw;
  return Math.abs(nowSeconds() - ts) <= spec.toleranceSec;
}

function encodedEqual(a: string, b: string, encoding: "hex" | "base64"): boolean {
  try {
    const ba = Buffer.from(a, encoding);
    const bb = Buffer.from(b, encoding);
    return ba.length > 0 && ba.length === bb.length && timingSafeEqual(ba, bb);
  } catch {
    return false;
  }
}

/**
 * webhook 서명 검증(provider-agnostic). 유효한 서명이 하나라도 맞으면 true.
 * 미서명/형식위반/시각초과/불일치 → false.
 */
export function verifyWebhookSignature(
  spec: WebhookSignatureSpec,
  rawBody: Uint8Array,
  headers: Headers,
  secret: string,
): boolean {
  const parsed = parseHeader(spec, headers);
  if (!parsed) return false;

  // replay 윈도우(타임스탬프가 있을 때만; "none"은 검사 안 함).
  if (spec.timestampSource !== "none") {
    if (parsed.timestampRaw === undefined) return false;
    if (!withinTolerance(spec, Number(parsed.timestampRaw))) return false;
  }

  // 서명 대상 페이로드 조립.
  const bodyStr = new TextDecoder().decode(rawBody);
  let signed = spec.payloadTemplate;
  if (signed.includes("{timestamp}")) {
    if (parsed.timestampRaw === undefined) return false;
    signed = signed.split("{timestamp}").join(parsed.timestampRaw);
  }
  if (signed.includes("{id}")) {
    const id = spec.idSource ? headers.get(spec.idSource.header) : null;
    if (id === null || id === "") return false;
    signed = signed.split("{id}").join(id);
  }
  signed = signed.split("{body}").join(bodyStr);

  const expected = createHmac(spec.algorithm, hmacKey(spec, secret))
    .update(signed)
    .digest(spec.encoding);
  return parsed.signatures.some((sig) => encodedEqual(expected, sig, spec.encoding));
}

File: apps/api-go/internal/modules/idp/signature.go
package idp

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
	"math"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// WebhookSignatureSpec는 설정형 webhook 서명 검증 spec이다(TS signature.ts 포팅).
// provider별 코드 대신 이 spec 하나로 ZITADEL / Stripe-style / Standard Webhooks / GitHub 등을
// 모두 검증한다(순수, crypto stdlib).
type WebhookSignatureSpec struct {
	// Header는 서명 헤더 이름. 예: "ZITADEL-Signature", "webhook-signature".
	Header string
	// HeaderFormat은 헤더 파싱 형식: "kv_t_v" | "scheme_hex" | "standard_webhooks".
	HeaderFormat string
	// TimestampSource는 타임스탬프 출처: "signature_header" | "separate_header" | "none".
	TimestampSource string
	// TimestampHeader는 TimestampSource="separate_header"일 때 타임스탬프 헤더 이름.
	TimestampHeader string
	// TimestampUnit은 타임스탬프 단위: "seconds" | "millis" | "none".
	TimestampUnit string
	// IDSource는 {id}를 쓰는 템플릿용 id 출처(예: Standard Webhooks "webhook-id").
	IDSource *IDSource
	// PayloadTemplate은 서명 대상 페이로드 템플릿. placeholders: {body} {timestamp} {id}.
	PayloadTemplate string
	// Algorithm은 "sha256".
	Algorithm string
	// Encoding은 헤더에 실린 서명 값의 인코딩: "hex" | "base64".
	Encoding string
	// SecretEncoding은 저장된 secret을 HMAC 키 바이트로 디코딩하는 방식: "raw" | "base64".
	SecretEncoding string
	// SecretPrefix는 디코딩 전 제거할 접두사(예: Standard Webhooks "whsec_").
	SecretPrefix string
	// ToleranceSec는 replay 허용 윈도우(초). too-old + far-future 모두 거부.
	ToleranceSec int
	// AllowMultipleSignatures는 다중 서명 허용(키 회전).
	AllowMultipleSignatures bool
}

// IDSource는 payload {id} 치환용 헤더 출처다.
type IDSource struct {
	Header string
}

// decimalRe는 엄격한 10진 정수(".0"/"e0"/leading-zero 거부)를 요구한다 —
// 타임스탬프를 raw 문자열로 보존하며 ZITADEL strconv.ParseInt과 동일하게 동작시킨다.
var decimalRe = regexp.MustCompile(`^[0-9]+$`)

// parsedSignature는 파싱된 (타임스탬프 raw, 서명값 목록)이다.
type parsedSignature struct {
	timestampRaw string // hasTimestamp=false면 무의미.
	hasTimestamp bool
	signatures   []string
}

// parseHeader는 헤더에서 (timestamp raw, 서명값 목록)을 추출한다. 형식 위반 시 ok=false.
func parseHeader(spec WebhookSignatureSpec, h http.Header) (parsedSignature, bool) {
	raw := h.Get(spec.Header)
	if raw == "" {
		return parsedSignature{}, false
	}

	var signatures []string
	inHeaderTsRaw := ""
	hasInHeaderTs := false

	switch spec.HeaderFormat {
	case "kv_t_v":
		// "t=<ts>,v1=<sig>[,v1=<sig>]"
		parts := map[string][]string{}
		for _, seg := range strings.Split(raw, ",") {
			i := strings.IndexByte(seg, '=')
			if i < 0 {
				continue
			}
			k := strings.TrimSpace(seg[:i])
			v := strings.TrimSpace(seg[i+1:])
			parts[k] = append(parts[k], v)
		}
		if ts := parts["t"]; len(ts) > 0 {
			// 엄격한 10진 정수만(".0"/"e0"/leading-zero 거부).
			if !decimalRe.MatchString(ts[0]) {
				return parsedSignature{}, false
			}
			inHeaderTsRaw = ts[0]
			hasInHeaderTs = true
		}
		for _, v := range parts["v1"] {
			if v != "" {
				signatures = append(signatures, v)
			}
		}
	case "scheme_hex":
		// "sha256=<hex>" — scheme label은 알고리즘과 일치해야 한다.
		eq := strings.IndexByte(raw, '=')
		if eq < 0 {
			return parsedSignature{}, false
		}
		if strings.ToLower(strings.TrimSpace(raw[:eq])) != spec.Algorithm {
			return parsedSignature{}, false
		}
		sig := strings.TrimSpace(raw[eq+1:])
		if sig == "" {
			return parsedSignature{}, false
		}
		signatures = []string{sig}
	default:
		// standard_webhooks: "v1,<base64>[ v1,<base64>]"
		for _, tok := range strings.Fields(raw) {
			comma := strings.IndexByte(tok, ',')
			if comma < 0 {
				continue
			}
			if tok[:comma] == "v1" && tok[comma+1:] != "" {
				signatures = append(signatures, tok[comma+1:])
			}
		}
	}

	if len(signatures) == 0 {
		return parsedSignature{}, false
	}
	if !spec.AllowMultipleSignatures {
		signatures = signatures[:1]
	}

	// 타임스탬프 출처 해석(raw 문자열로 보존).
	timestampRaw := ""
	hasTimestamp := false
	switch spec.TimestampSource {
	case "signature_header":
		if !hasInHeaderTs {
			return parsedSignature{}, false
		}
		timestampRaw = inHeaderTsRaw
		hasTimestamp = true
	case "separate_header":
		tv := ""
		if spec.TimestampHeader != "" {
			tv = h.Get(spec.TimestampHeader)
		}
		if !decimalRe.MatchString(tv) {
			return parsedSignature{}, false
		}
		timestampRaw = tv
		hasTimestamp = true
	}
	return parsedSignature{timestampRaw: timestampRaw, hasTimestamp: hasTimestamp, signatures: signatures}, true
}

// hmacKey는 spec.SecretEncoding/SecretPrefix에 따라 HMAC 키 바이트를 만든다.
func hmacKey(spec WebhookSignatureSpec, secret string) []byte {
	s := secret
	if spec.SecretPrefix != "" && strings.HasPrefix(s, spec.SecretPrefix) {
		s = s[len(spec.SecretPrefix):]
	}
	if spec.SecretEncoding == "base64" {
		if b, err := base64.StdEncoding.DecodeString(s); err == nil {
			return b
		}
		if b, err := base64.RawStdEncoding.DecodeString(s); err == nil {
			return b
		}
		return nil // 디코딩 불가 → 빈 키(검증 실패로 이어짐).
	}
	return []byte(s)
}

// withinTolerance는 타임스탬프(초)를 toleranceSec 양방향 윈도우로 검사한다.
func withinTolerance(spec WebhookSignatureSpec, tsRaw float64, now func() time.Time) bool {
	ts := tsRaw
	if spec.TimestampUnit == "millis" {
		ts = math.Floor(tsRaw / 1000)
	}
	return math.Abs(float64(now().Unix())-ts) <= float64(spec.ToleranceSec)
}

// decodeSig는 헤더 서명값을 인코딩에 맞게 바이트로 디코딩한다. 실패 시 ok=false.
func decodeSig(s, encoding string) ([]byte, bool) {
	if encoding == "hex" {
		b, err := hex.DecodeString(s)
		if err != nil {
			return nil, false
		}
		return b, true
	}
	// base64(패딩/무패딩 모두 허용 — Node Buffer.from(...,"base64")의 관대함 근사).
	if b, err := base64.StdEncoding.DecodeString(s); err == nil {
		return b, true
	}
	if b, err := base64.RawStdEncoding.DecodeString(s); err == nil {
		return b, true
	}
	return nil, false
}

// VerifyWebhookSignature는 raw webhook 본문을 선언적 서명 spec으로 검증한다.
// 유효한 서명이 하나라도 맞으면 true. 미서명/형식위반/시각초과/불일치 → false.
// now는 replay 윈도우 검사용 주입 클록(프로덕션은 time.Now).
func VerifyWebhookSignature(spec WebhookSignatureSpec, rawBody []byte, h http.Header, secret string, now func() time.Time) bool {
	parsed, ok := parseHeader(spec, h)
	if !ok {
		return false
	}

	// replay 윈도우(타임스탬프가 있을 때만; "none"은 검사 안 함).
	if spec.TimestampSource != "none" {
		if !parsed.hasTimestamp {
			return false
		}
		tsNum, err := strconv.ParseFloat(parsed.timestampRaw, 64)
		if err != nil {
			return false
		}
		if !withinTolerance(spec, tsNum, now) {
			return false
		}
	}

	// 서명 대상 페이로드 조립(타임스탬프는 raw 문자열 그대로).
	signed := spec.PayloadTemplate
	if strings.Contains(signed, "{timestamp}") {
		if !parsed.hasTimestamp {
			return false
		}
		signed = strings.ReplaceAll(signed, "{timestamp}", parsed.timestampRaw)
	}
	if strings.Contains(signed, "{id}") {
		id := ""
		if spec.IDSource != nil {
			id = h.Get(spec.IDSource.Header)
		}
		if id == "" {
			return false
		}
		signed = strings.ReplaceAll(signed, "{id}", id)
	}
	signed = strings.ReplaceAll(signed, "{body}", string(rawBody))

	mac := hmac.New(sha256.New, hmacKey(spec, secret))
	mac.Write([]byte(signed))
	sum := mac.Sum(nil)

	// decode-then-constant-time compare(비어있지 않고 길이 같을 때만).
	for _, sig := range parsed.signatures {
		cand, ok := decodeSig(sig, spec.Encoding)
		if !ok {
			continue
		}
		if len(cand) == len(sum) && subtle.ConstantTimeCompare(cand, sum) == 1 {
			return true
		}
	}
	return false
}
